app/vlselect/traces/query/query.go

```diff
@@ -0,0 +1,404 @@
+				}
+			}
+
+			rowsLock.Lock()
+			rows = append(rows, &Row{
+				Timestamp: timestamp,
+				Fields:    fields,
+			})
+			rowsLock.Unlock()
+		}
+	}
+
+	if err = vlstorage.RunQuery(ctxWithCancel, cp.TenantIDs, q, writeBlock); err != nil {
+		return nil, nil, err
+	}
+	if missingTimeColumn.Load() {
+		return nil, nil, fmt.Errorf("missing _time column in the result for the query [%s]", q)
+	}
+	return traceIDs, rows, nil
+}
+
+// getTraceIDList returns traceIDs according to the search params.
+// It also returns the earliest start time of these traces, to help reducing the time range for spans search.
+func getTraceIDList(ctx context.Context, cp *CommonParams, param *TraceQueryParam) ([]string, time.Time, error) {
+	currentTime := time.Now()
+	// query: * AND <filter> | last 1 by (_time) partition by (trace_id) | fields _time, trace_id | sort by (_time) desc
+	qStr := "*"
+	if param.ServiceName != "" {
+		qStr += fmt.Sprintf("AND _stream:{"+pb.ResourceAttrServiceName+"=\"%s\"} ", param.ServiceName)
+	}
+	if param.SpanName != "" {
+		qStr += fmt.Sprintf("AND _stream:{"+pb.NameField+"=\"%s\"} ", param.SpanName)
+	}
+	if len(param.Attributes) > 0 {
+		for k, v := range param.Attributes {
+			qStr += fmt.Sprintf(`AND "%s":="%s" `, k, v)
+		}
+	}
+	if param.DurationMin > 0 {
+		qStr += fmt.Sprintf("AND "+pb.DurationField+":>%d ", param.DurationMin.Nanoseconds())
+	}
+	if param.DurationMax > 0 {
+		qStr += fmt.Sprintf("AND duration:<%d ", param.DurationMax.Nanoseconds())
+	}
+	qStr += " | last 1 by (_time) partition by (" + pb.TraceIDField + ") | fields _time, " + pb.TraceIDField + " | sort by (_time) desc"
+
+	q, err := logstorage.ParseQueryAtTimestamp(qStr, currentTime.UnixNano())
+	if err != nil {
+		return nil, time.Time{}, fmt.Errorf("cannot parse query [%s]: %s", qStr, err)
+	}
+	q.AddPipeLimit(uint64(param.Limit))
+
+	traceIDs, maxStartTime, err := findTraceIDsSplitTimeRange(ctx, q, cp, param.StartTimeMin, param.StartTimeMax, param.Limit)
+	if err != nil {
+		return nil, maxStartTime, err
+	}
+
+	return traceIDs, maxStartTime, nil
+}
+
+// findTraceIDsSplitTimeRange try to search from the nearest time range of the end time.
+// if the result already met requirement of `limit`, return.
+// otherwise, amplify the time range to 5x and search again, until the start time exceed the input.
+func findTraceIDsSplitTimeRange(ctx context.Context, q *logstorage.Query, cp *CommonParams, startTime, endTime time.Time, limit int) ([]string, time.Time, error) {
+	currentTime := time.Now()
+
+	step := time.Minute
+	currentStartTime := endTime.Add(-step)
+
+	traceIDList := make([]string, 0, limit)
+	maxStartTimeStr := endTime.Format(time.RFC3339)
+
+	writeBlock := func(_ uint, db *logstorage.DataBlock) {
+		columns := db.Columns
+		for i := range columns {
+			if columns[i].Name == "trace_id" {
+				traceIDList = append(traceIDList, columns[i].Values...)
```

Comment:
Shouldn't we use `strings.Clone` here and on line 361 as well? Since you used it in other writeBlock functions, I assume we should use it too.